### PR TITLE
Installation type: multiple operating systems

### PR DESCRIPTION
**Pull request by jpnurmi**
Take into account that one may have the same OS installed multiple times. In this case, select:

> This computer currently has multiple operating systems on it.

instead of for example:

> This computer currently has Ubuntu 22.04 and Ubuntu 22.04 on it.

Fixes: #1046

### CODE REVIEW

File: packages/ubuntu_desktop_installer/lib/pages/installation_type/installation_type_page.dart [hunk 1]
```diff
@@ -51,7 +51,9 @@ class _InstallationTypePageState extends State<InstallationTypePage> {
       case 1:
         return lang.installationTypeOSDetected(os.single.long);
       case 2:
+        if (os.hasDuplicates) continue duplicates;
```
Comment (CarlosNihelton) [inline packages/ubuntu_desktop_installer/lib/pages/installation_type/installation_type_page.dart:54]: Honest question: do you think this jump reads better than leaving it fall through the default case?

```dart
  if (!os.hasDuplicates) {
    return lang.installationTypeDualOSDetected(os.first.long, os.last.long);
  }

default:
    return lang.installationTypeMultiOSDetected;
```

**Review by CarlosNihelton — APPROVED** [review]
One minor question, otherwise +1 !

File: packages/ubuntu_desktop_installer/lib/pages/installation_type/installation_type_page.dart (hunk 1 above)
Comment (jpnurmi) [reply packages/ubuntu_desktop_installer/lib/pages/installation_type/installation_type_page.dart:54]: That was actually my first attempt too but the Dart analyzer doesn't allow falling through unless the case is empty ([switch_case_completes_normally](https://dart.dev/tools/diagnostic-messages?utm_source=dartdev&utm_medium=redir&utm_id=diagcode&utm_content=switch_case_completes_normally#switch_case_completes_normally)).

Do you like something like this more?
```diff
diff --git a/packages/ubuntu_desktop_installer/lib/pages/installation_type/installation_type_page.dart b/packages/ubuntu_desktop_installer/lib/pages/installation_type/installation_type_pag>
index 85a143f9..01c76bb0 100644
--- a/packages/ubuntu_desktop_installer/lib/pages/installation_type/installation_type_page.dart
+++ b/packages/ubuntu_desktop_installer/lib/pages/installation_type/installation_type_page.dart
@@ -51,9 +51,9 @@ class _InstallationTypePageState extends State<InstallationTypePage> {
       case 1:
         return lang.installationTypeOSDetected(os.single.long);
       case 2:
-        if (os.hasDuplicates) continue duplicates;
-        return lang.installationTypeDualOSDetected(os.first.long, os.last.long);
-      duplicates:
+        return os.hasDuplicates
+            ? lang.installationTypeMultiOSDetected
+            : lang.installationTypeDualOSDetected(os.first.long, os.last.long);
       default:
         return lang.installationTypeMultiOSDetected;
     }
@@ -68,10 +68,10 @@ class _InstallationTypePageState extends State<InstallationTypePage> {
       case 1:
         return lang.installationTypeAlongside(product, os.single.long);
       case 2:
-        if (os.hasDuplicates) continue duplicates;
-        return lang.installationTypeAlongsideDual(
-            product, os.first.long, os.last.long);
-      duplicates:
+        return os.hasDuplicates
+            ? lang.installationTypeAlongsideMulti(product)
+            : lang.installationTypeAlongsideDual(
+                product, os.first.long, os.last.long);
       default:
         return lang.installationTypeAlongsideMulti(product);
     }
```

Comment (CarlosNihelton) [reply packages/ubuntu_desktop_installer/lib/pages/installation_type/installation_type_page.dart:54]: Honestly I do.

Comment (jpnurmi) [reply packages/ubuntu_desktop_installer/lib/pages/installation_type/installation_type_page.dart:54]: how the heck can we make this pretty :D
```dart
return os.length > 2 || os.hasDuplicates
    ? lang.installationTypeMultiOSDetected
    : os.length == 2
        ? lang.installationTypeDualOSDetected(os.first.long, os.last.long)
        : os.length == 1
            ? lang.installationTypeOSDetected(os.single.long)
            : lang.installationTypeNoOSDetected;
```

Comment (jpnurmi) [reply packages/ubuntu_desktop_installer/lib/pages/installation_type/installation_type_page.dart:54]: meh...
```dart
final count = os.hasDuplicates ? -1 : os.length;
switch (count) {
  case 0:
    return lang.installationTypeNoOSDetected;
  case 1:
    return lang.installationTypeOSDetected(os.single.long);
  case 2:
    return lang.installationTypeDualOSDetected(os.first.long, os.last.long);
  default:
    return lang.installationTypeMultiOSDetected;
}
```

Comment (CarlosNihelton) [reply packages/ubuntu_desktop_installer/lib/pages/installation_type/installation_type_page.dart]: That would work as well. I like the ternary as is, though.

Comment (jpnurmi) [reply packages/ubuntu_desktop_installer/lib/pages/installation_type/installation_type_page.dart]: Thanks, it looks more readable than the first proposal :)

**Comment by corradoventu** [comment]
Still wrorg in subiquity 573 canary ISO dated 2022-08-13 09:33 in the screen installation-type i see:
The computer currently has Ubuntu 22.04 on it... 
while I have multiple OS: 
corrado@corrado-n6-kk-0811:~$ sudo update-grub
[sudo] password for corrado: 
Sourcing file `/etc/default/grub'
Sourcing file `/etc/default/grub.d/init-select.cfg'
Generating grub configuration file ...
Found background image: aPartn6.png
Found linux image: /boot/vmlinuz-5.15.0-27-generic
Found initrd image: /boot/initrd.img-5.15.0-27-generic
Memtest86+ needs a 16-bit boot, that is not available on EFI, exiting
Warning: os-prober will be executed to detect other bootable partitions.
Its output will be used to detect bootable binaries on them and create new boot entries.
Found Ubuntu 22.04.1 LTS (22.04) on /dev/nvme0n1p2
Found Ubuntu Kinetic Kudu (development branch) (22.10) on /dev/nvme0n1p3
Found Ubuntu Kinetic Kudu (development branch) (22.10) on /dev/nvme0n1p4
Found Ubuntu 20.04.4 LTS (20.04) on /dev/nvme0n1p5
Found Ubuntu 22.04 LTS (22.04) on /dev/sda3
Adding boot menu entry for UEFI Firmware Settings ...
done
corrado@corrado-n6-kk-0811:~$

**Comment by jpnurmi** [comment]
Try https://github.com/canonical/ubuntu-desktop-installer/pull/1044#issuecomment-1214715109:
```
snap refresh --channel=candidate ubuntu-desktop-installer
```

**Comment by corradoventu** [comment]
Still wrong in subiquity 588 canary ISO dated 2022-08-16 in the screen installation-type i see:
The computer currently has no detected...
while I have multiple OS:
```
corrado@corrado-n6-kk-0811:~$ sudo update-grub
[sudo] password for corrado: 
Sourcing file `/etc/default/grub'
Sourcing file `/etc/default/grub.d/init-select.cfg'
Generating grub configuration file ...
Found background image: aPartn6.png
Found linux image: /boot/vmlinuz-5.15.0-27-generic
Found initrd image: /boot/initrd.img-5.15.0-27-generic
Memtest86+ needs a 16-bit boot, that is not available on EFI, exiting
Warning: os-prober will be executed to detect other bootable partitions.
Its output will be used to detect bootable binaries on them and create new boot entries.
Found Ubuntu 22.04.1 LTS (22.04) on /dev/nvme0n1p2
Found Ubuntu Kinetic Kudu (development branch) (22.10) on /dev/nvme0n1p3
Found Ubuntu Kinetic Kudu (development branch) (22.10) on /dev/nvme0n1p4
Found Ubuntu 20.04.4 LTS (20.04) on /dev/nvme0n1p5
Found Ubuntu 22.04 LTS (22.04) on /dev/sda3
Adding boot menu entry for UEFI Firmware Settings ...
done
corrado@corrado-n6-kk-0811:~$ 
```
see bug https://bugs.launchpad.net/subiquity/+bug/1986676 for log and screenshots

**Comment by corradoventu** [comment]
Still same problem with subiquity 598: only one OS detected. OsProber not called for nvme disk?
2022-08-23 14:38:17.633508 DEBUG disk_storage: Update storage: StorageResponseV2(status: ProbeStatus.DONE, errorReport: null, disks: [Disk(id: disk-nvme0n1, label: KINGSTON_SKC2000M8250G_50026B72823D1475, type: local disk, size: 250059350016, usageLabels: [], partitions: [PartitionOrGap.partition(size: 268435456, number: 1, preserve: true, wipe: null, annotations: [existing, unused ESP, already formatted as vfat], mount: null, format: vfat, grubDevice: false, boot: true, os: null, offset: 1048576, estimatedMinSize: -1, resize: null, path: /dev/nvme0n1p1), PartitionOrGap.partition(size: 42949672960, number: 2, preserve: true, wipe: null, annotations: [existing, already formatted as ext4, not mounted], mount: null, format: ext4, grubDevice: false, boot: false, os: null, offset: 269484032, estimatedMinSize: 13871611904, resize: null, path: /dev/nvme0n1p2), PartitionOrGap.partition(size: 42949672960, number: 3, preserve: true, wipe: null, annotations: [existing, already formatted as ext4, not mounted], mount: null, format: ext4, grubDevice: false, boot: false, os: null, offset: 43219156992, estimatedMinSize: 21851275264, resize: null, path: /dev/nvme0n1p3), PartitionOrGap.partition(size: 42949672960, number: 4, preserve: true, wipe: null, annotations: [existing, already formatted as ext4, not mounted], mount: null, format: ext4, grubDevice: false, boot: false, os: null, offset: 86168829952, estimatedMinSize: 25813843968, resize: null, path: /dev/nvme0n1p4), PartitionOrGap.partition(size: 42949672960, number: 5, preserve: true, wipe: null, annotations: [existing, already formatted as ext4, not mounted], mount: null, format: ext4, grubDevice: false, boot: false, os: null, offset: 129118502912, estimatedMinSize: 9787408384, resize: null, path: /dev/nvme0n1p5), PartitionOrGap.partition(size: 42949672960, number: 6, preserve: true, wipe: null, annotations: [existing, already formatted as ext4, not mounted], mount: null, format: ext4, grubDevice: false, boot: false, os: null, offset: 172068175872, estimatedMinSize: 20633878528, resize: null, path: /dev/nvme0n1p6), PartitionOrGap.partition(size: 35041312768, number: 7, preserve: true, wipe: null, annotations: [existing, already formatted as ext4, not mounted], mount: null, format: ext4, grubDevice: false, boot: false, os: null, offset: 215017848832, estimatedMinSize: 468713472, resize: null, path: /dev/nvme0n1p7)], okForGuided: true, ptable: gpt, preserve: true, path: /dev/nvme0n1, bootDevice: false, model: null, vendor: null), Disk(id: disk-sda, label: CT500MX500SSD1_21122DB99114, type: local disk, size: 500107862016, usageLabels: [], partitions: [PartitionOrGap.partition(size: 268435456, number: 1, preserve: true, wipe: null, annotations: [existing, already formatted as vfat, not mounted], mount: null, format: vfat, grubDevice: false, boot: false, os: null, offset: 1048576, estimatedMinSize: -1, resize: null, path: /dev/sda1), PartitionOrGap.partition(size: 8589934592, number: 2, preserve: true, wipe: null, annotations: [existing, already formatted as swap, unused], mount: null, format: swap, grubDevice: false, boot: false, os: null, offset: 269484032, estimatedMinSize: 1048576, resize: null, path: /dev/sda2), PartitionOrGap.partition(size: 34359738368, number: 3, preserve: true, wipe: null, annotations: [existing, already formatted as ext4, not mounted], mount: null, format: ext4, grubDevice: false, boot: false, os: OsProber(long: Ubuntu 22.04 LTS, label: Ubuntu5, type: linux, subpath: null, version: 22.04), offset: 8859418624, estimatedMinSize: 9489612800, resize: null, path: /dev/sda3), PartitionOrGap.partition(size: 34359738368, number: 4, preserve: true, wipe: null, annotations: [existing, already formatted as ext4, not mounted], mount: null, format: ext4, grubDevice: false, boot: false, os: null, offset: 43219156992, estimatedMinSize: 469762048, resize: null, path: /dev/sda4), PartitionOrGap.partition(size: 422528942080, number: 5, preserve: true, wipe: null, annotations: [existing, already formatted as ext4, not mounted], mount: null, format: ext4, grubDevice: false, boot: false, os: null, offset: 77578895360, estimatedMinSize: 2083520512, resize: null, path: /dev/sda5)], okForGuided: true, ptable: gpt, preserve: true, path: /dev/sda, bootDevice: false, model: CT500MX500SSD1, vendor: null), Disk(id: disk-sdb, label: CT500MX500SSD1_21122DA80425, type: local disk, size: 500107862016, usageLabels: [], partitions: [PartitionOrGap.partition(size: 34359738368, number: 1, preserve: true, wipe: null, annotations: [existing, already formatted as ext4, not mounted], mount: null, format: ext4, grubDevice: false, boot: false, os: null, offset: 1048576, estimatedMinSize: 469762048, resize: null, path: /dev/sdb1), PartitionOrGap.partition(size: 34359738368, number: 2, preserve: true, wipe: null, annotations: [existing, already formatted as ext4, not mounted], mount: null, format: ext4, grubDevice: false, boot: false, os: null, offset: 34360786944, estimatedMinSize: 469762048, resize: null, path: /dev/sdb2), PartitionOrGap.partition(size: 431387312128, number: 3, preserve: true, wipe: null, annotations: [existing, already formatted as ext4, not mounted], mount: null, format: ext4, grubDevice: false, boot: false, os: null, offset: 68720525312, estimatedMinSize: 197694324736, resize: null, path: /dev/sdb3)], okForGuided: true, ptable: gpt, preserve: true, path: /dev/sdb, bootDevice: false, model: CT500MX500SSD1, vendor: null), Disk(id: disk-sdd, label: Generic_USB_SD_Reader-0:0, type: local disk, size: 0, usageLabels: [], partitions: [], okForGuided: true, ptable: null, preserve: true, path: /dev/sdd, bootDevice: false, model: USB SD Reader, vendor: Generic), Disk(id: disk-sde, label: Generic_USB_CF_Reader-0:1, type: local disk, size: 0, usageLabels: [], partitions: [], okForGuided: true, ptable: null, preserve: true, path: /dev/sde, bootDevice: false, model: USB CF Reader, vendor: Generic), Disk(id: disk-sdf, label: Generic_USB_SM_Reader-0:2, type: local disk, size: 0, usageLabels: [], partitions: [], okForGuided: true, ptable: null, preserve: true, path: /dev/sdf, bootDevice: false, model: USB SM Reader, vendor: Generic), Disk(id: disk-sdg, label: Generic_USB_MS_Reader-0:3, type: local disk, size: 0, usageLabels: [], partitions: [], okForGuided: true, ptable: null, preserve: true, path: /dev/sdg, bootDevice: false, model: USB MS Reader, vendor: Generic)], needRoot: true, needBoot: true, installMinimumSize: 18484297728)

**Pull request merged**